daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -1503,7 +1718,92 @@ func (devices *DeviceSet) DeleteDevice(hash string) error {
 	devices.Lock()
 	defer devices.Unlock()
 
-	return devices.deleteDevice(info)
+	// If mountcount is not zero, that means devices is still in use
+	// or has not been Put() properly. Fail device deletion.
+
+	if info.mountCount != 0 {
+		return fmt.Errorf("devmapper: Can't delete device %v as it is still mounted. mntCount=%v", info.Hash, info.mountCount)
+	}
+
+	return devices.deleteDevice(info, syncDelete)
+}
+
+// Either sleep for 30 seconds or wait for command from parent.
+// It returns true if one should continue to loop. Returns false if
+// parent sent a message on channel indicating stop.
+func (devices *DeviceSet) waitSleepCleanupThread() bool {
+	// Create a timer of 30 seconds.
+	timer := time.NewTimer(time.Second * 30)
+
+	select {
+	case <-timer.C:
+		return true
+	case <-devices.deletionWorkerChan:
+		timer.Stop()
+	}
+	return false
+}
+
+// Don't call this function directly. Should be scheduled with a call to
+// scheduleDefDeletion()
+func (devices *DeviceSet) cleanupDeletedDevicesThread() {
+	devices.Lock()
+	for {
+		logrus.Debugf("devmapper: Calling cleanupDeletedDevices().\n")
+		err := devices.cleanupDeletedDevices()
+		if err != nil {
+			logrus.Errorf("devmapper: cleanupDeletedDevices() failed.")
```

Comment:
given that this may fail, and the `devices.deletionWorkerRunning` would not be reset to `false`, then this thread would never be respawned until daemon restart, correct?